rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -865,63 +948,128 @@ void main_window::HandlePupInstallation(QString file_path)
 		critical(tr("Firmware installation failed: The selected firmware file couldn't be opened."));
 		return;
 	}
 
-	if (pup_f.size() < sizeof(PUPHeader))
+	pup_object pup(std::move(pup_f));
+
+	switch (pup.operator pup_error())
 	{
-		gui_log.error("Too small PUP file: %llu", pup_f.size());
+	case pup_error::header_read:
+	{
+		gui_log.error("%s", pup.get_formatted_error());
 		critical(tr("Firmware installation failed: The provided file is empty."));
 		return;
 	}
-
-	struct PUPHeader header = {};
-	pup_f.seek(0);
-	pup_f.read(header);
-
-	if (header.header_length + header.data_length != pup_f.size())
+	case pup_error::header_magic:
 	{
-		gui_log.error("Firmware size mismatch, expected: %llu + %llu, actual: %llu", header.header_length, header.data_length, pup_f.size());
+		gui_log.error("Error while installing firmware: provided file is not a PUP file.");
+		critical(tr("Firmware installation failed: The provided file is not a PUP file."));
+		return;
+	}
+	case pup_error::expected_size:
+	{
+		gui_log.error("%s", pup.get_formatted_error());
 		critical(tr("Firmware installation failed: The provided file is incomplete. Try redownloading it."));
 		return;
 	}
-
-	pup_object pup(pup_f);
-	if (!pup)
+	case pup_error::header_file_count:
+	case pup_error::file_entries:
+	default:
 	{
-		gui_log.error("Error while installing firmware: PUP file is invalid.");
+		std::string error = "Error while installing firmware: PUP file is invalid.";
+
+		if (!pup.get_formatted_error().empty())
+		{
+			fmt::append(error, "\n%s", pup.get_formatted_error());
+		}
+
+		gui_log.error("%s", error);
 		critical(tr("Firmware installation failed: The provided file is corrupted."));
 		return;
 	}
-
-	if (!pup.validate_hashes())
+	case pup_error::hash_mismatch:
 	{
 		gui_log.error("Error while installing firmware: Hash check failed.");
 		critical(tr("Firmware installation failed: The provided file's contents are corrupted."));
 		return;
 	}
+	case pup_error::ok: break;
+	}
 
 	fs::file update_files_f = pup.get_file(0x300);
+
+	if (!update_files_f)
+	{
+		gui_log.error("Error while installing firmware: Couldn't find installation packages database.");
+		critical(tr("Firmware installation failed: The provided file's contents are corrupted."));
+		return;
+	}
+
 	tar_object update_files(update_files_f);
+
+	if (!dir_path.isEmpty())
+	{
+		// Extract only mode, extract direct TAR entries to a user directory
+
+		if (!vfs::mount("/pup_extract", sstr(dir_path) + '/'))
+		{
+			gui_log.error("Error while extracting firmware: Failed to mount '%s'", sstr(dir_path));
+			critical(tr("Firmware extraction failed: VFS mounting failed."));
+			return;	
+		}
+
+		if (!update_files.extract("/pup_extract"))
+		{
+			gui_log.error("Error while installing firmware: TAR contents are invalid.");
+			critical(tr("Firmware installation failed: Firmware contents could not be extracted."));
+		}
+
+		gui_log.success("Extracted PUP file to %s!", sstr(dir_path));
```

Comment:
This may lead to confusion (people already thought that some ! was part of some path)
```suggestion
		gui_log.success("Extracted PUP file to %s", sstr(dir_path));
```